---
title: "Metabarcoding of mosquito species from french continental and oversea territories"
author: "Hans Schrieke"
date: "26/07/2019"
output: 
  html_document:
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# I : From raw sequencing data to ASV table (dada2)

For this study, we have 3 runs of data and we make a dada2 process run by run and merge the different output tables in a unique.  

For Dada2 step, I just show the processing of one of my runs (run 1). The difference with the other runs is the choosen parameters for filter and trim step. 

If you have only one run, you can ignore the "merge run tables in a unique" step.

If you have already your sequence, taxonomic tables and your metadata file, you can go to the second part "Decontamination and normalization" for load your files and proceed to decontamination and analysis. 

## Set directory and load packages
```{r}
# Set your working directory 
path <- "D:/stage/data/dada2/run1"
setwd(path)

# Load packages 
library(dada2); packageVersion("dada2")
```

## Prepare sequences 
```{r}
  # Store forward and reverse reads 
  forward <- sort(list.files(path, pattern="_R1_001.fastq.gz", full.names = TRUE))
  reverse <- sort(list.files(path, pattern="_R2_001.fastq.gz", full.names = TRUE))
  
  # Store sample names 
  sample.names <- sapply(strsplit(basename(forward), "_"), `[`, 1)
```

## Inspect read quality profiles
```{r}
plotQualityProfile(forward[1:2])

plotQualityProfile(reverse[1:2])
```


## Filter and trim
```{r}
 
  
  # Prepare folds and names of filter reads
  filtForward <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
  filtReverse <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
  
  # Filter and trim process
  out <- filterAndTrim(forward, filtForward, reverse, filtReverse, truncLen=c(245,235),
                            maxN=0, maxEE=c(2,4), truncQ=2, rm.phix=TRUE,
                            compress=TRUE, multithread=TRUE) # multithread=FALSE on windows
  
  # # Parameters for run2: 
  #    out <- filterAndTrim(forward, filtForward, reverse, filtReverse, truncLen=c(245,235),
  #                       maxN=0, maxEE=c(2,4), truncQ=2, rm.phix=TRUE,
  #                       compress=TRUE, multithread=TRUE)
  #                      
  # # Parameters for run3:
  #      out <- filterAndTrim(forward, filtForward, reverse, filtReverse, truncLen=c(245,235),
  #                       maxN=0, maxEE=c(2,4), truncQ=2, rm.phix=TRUE,
  #                       compress=TRUE, multithread=TRUE)
  
  
  # Check output
  head(out)
  
```


## Learn error rates

```{r}
  # Learn errors
  errForward <- learnErrors(filtForward, multithread=TRUE)
  errReverse <- learnErrors(filtReverse, multithread=TRUE)

```

## Dereplication, sample inference and merge

```{r}
  # Dereplication
  derepForward <- derepFastq(filtForward, verbose=TRUE)
  derepReverse <- derepFastq(filtReverse, verbose=TRUE)
  
  #Name the derep-class objects by the sample names
  names(derepForward) <- sample.names
  names(derepReverse) <- sample.names

  # Sample inference
  dadaForward <- dada(derepForward, err=errForward, multithread=TRUE)
  dadaReverse <- dada(derepReverse, err=errReverse, multithread=TRUE)
  
  # Merge
  mergers <- mergePairs(dadaForward, derepForward, dadaReverse, derepReverse, verbose=TRUE)
  
  # Check merge
  head(mergers[[1]])
```

## Remove chimera and write ASV table 
```{r}
  # Construct table 
  seqtab <- makeSequenceTable(mergers)
  
  # Construct table without chimera
  seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
  
  # Percentage of removing chimera
  print(paste0("Percentage of chimera: ",100-(sum(seqtab.nochim)/sum(seqtab)*100)))
  
  # Write tables to disk 
  write.table(t(seqtab),  "seqtab1.csv", sep=";", dec=",")
  write.table(t(seqtab.nochim),  "seqtabnochim1.csv", sep=";", dec=",")
  saveRDS(seqtab.nochim, file="seqtab1.rds")
  
  
```

## A few stats 

```{r}
  # Sum function
  getN <- function(x) sum(getUniques(x))
  
  # Create table with sum of dada forward reads, dada reverse reads, merged reads and count of reads
  track <- cbind(out, sapply(dadaForward, getN), sapply(dadaReverse, getN), 
                  sapply(mergers, getN),rowSums(seqtab.nochim))
  
  # Rename columns and rows
  colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
  rownames(track) <- sample.names
  
  # Check track 
  head(track)
  
  # Write track on disk 
  write.table(track,"stats1.csv",sep=";",dec=",") 
  saveRDS(track, file="stats1.rds")
```


## Merge run tables in a unique table 
```{r}
  # Set directory to output of run1, run2, run3
  path_run1 <- "D:/stage/data/dada2/run1/output"
  path_run2 <- "D:/stage/data/dada2/run2/output"
  path_run3 <- "D:/stage/data/dada2/run3/output"
  
  # Load seqtab of the different runs
  st1 <- readRDS("D:/stage/data/dada2/run1/output/seqtab1.rds")
  st2 <- readRDS("D:/stage/data/dada2/run2/output/seqtab2.rds")
  st3 <- readRDS("D:/stage/data/dada2/run3/output/seqtab3.rds")
  
  # Invert and convert tables to data frame
  st1 <- t(st1)
  st2 <- t(st2)
  st3 <- t(st3)

  st1 <- as.data.frame(st1)
  st2 <- as.data.frame(st2)
  st3 <- as.data.frame(st3)
  
  #Change Undetermined sequence names for each run
  names(st1)[97] <- "Undetermined1"
  names(st2)[80] <- "Undetermined2"
  names(st3)[78] <- "Undetermined3"
  
  # Invert and convert tables to matrix
  st1 <- t(st1)
  st2 <- t(st2)
  st3 <- t(st3)

  st1 <- as.matrix(st1)
  st2 <- as.matrix(st2)
  st3 <- as.matrix(st3)
  
  
  # Merge tables 
  st.all <- mergeSequenceTables(st1, st2, st3)
  st.all.good <- t(st.all)
  
  # Write final table to disk
  write.table(st.all.good, "seqtabnochim.csv", sep=";", dec=",")

```


## Taxonomic assignment and write taxa table 

```{r}
  # Assign taxonomy with silva132 database reference
  taxa <- assignTaxonomy(st.all, file.path(path, "silva_nr_v132_train_set.fa.gz"), 
                          multithread=TRUE, minBoot=80)
                          
  # Add species-level annotation to the taxonomic table
  taxa <- addSpecies(taxa, file.path(path, "silva_species_assignment_v132.fa.gz"))
  
  # Write track on disk 
  write.table(taxa,"taxafinal.csv",sep=";",dec=",")
  saveRDS(taxa, file="taxa1.rds")
```

```{r}
head(taxa)
```


## Optional : phylogenetic tree`

```{r}
  # Load packages
  library("DECIPHER")
  library("phangorn")
  
  otu <- read.csv("seqtabnochim.csv", sep=";", dec=",")
  otu <- as.matrix(t(otu))

  seqs <- getSequences(otu)
  names(seqs) <- seqs # This propagates to the tip labels of the tree
  alignment <- AlignSeqs(DNAStringSet(seqs), anchor=NA,verbose=FALSE)



  phangAlign <- phyDat(as(alignment, "matrix"), type="DNA")
  dm <- dist.ml(phangAlign)



  treeNJ <- NJ(dm) # Note, tip order != sequence order
  fit = pml(treeNJ, data=phangAlign)
  fitGTR <- update(fit, k=4, inv=0.2)
  fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
                    rearrangement = "stochastic", control = pml.control(trace = 0))
  detach("package:phangorn", unload=TRUE)

```

## Save session
```{r}
library(session); packageVersion("session")
save.session("metabarcoding.Rda")
```

# II : Decontamination and normalization 

## Set path and load packages
```{r}
path = "D:/stage/data/runs_new2/"
setwd(path)

dir.create("decontam_script")
path_main <- "D:/stage/data/runs_new2/decontam_script/"# folder for plot
setwd(path_main)

dir.create("1.raw_data")
path_raw <- "D:/stage/data/runs_new2/decontam_script/1.raw_data"

dir.create("2.decontam")
path_decontam <- "D:/stage/data/runs_new2/decontam_script/2.decontam"

dir.create("3.decontam2")
path_decontam2 <- "D:/stage/data/runs_new2/decontam/script/3.decontam2"

library(dada2); packageVersion("dada2")
library(phyloseq); packageVersion("phyloseq")
library(ggplot2); packageVersion("ggplot2")
library(decontam); packageVersion("decontam")
library(DESeq2); packageVersion("DESeq2")

theme_set(theme_gray()) #set ggplot2 graphic theme
```

## Import and prepare data 
```{r}
setwd(path)
asv_tab <- read.csv("seqtabnochimcor.csv", sep=";", dec=",")
asv_tax <- read.csv("taxafinal.csv", sep=";", dec=",")
metadata <- read.csv("metadata_runs_update_02_07_19.csv", sep=",", row.names = 1)
load("tree.Rdata")


setwd(path_raw)

# giving our seq headers more manageable names (ASV_1, ASV_2...)
asv_seqs <- rownames(asv_tab)
asv_headers <- vector(dim(asv_tab)[1], mode="character")
asv_headers2 <- vector(dim(asv_tab)[1], mode="character")

for (i in 1:dim(asv_tab)[1]) {
  asv_headers[i] <- paste(">ASV", i, sep="_")
  asv_headers2[i] <- paste("ASV", i, sep="_")
}

# making and writing out a fasta of our final ASV seqs:
asv_fasta_standard <- c(rbind(asv_headers, asv_seqs))
asv_fasta <- cbind(asv_headers2, asv_seqs)
write(asv_fasta_standard, "ASVs_standard.fa")
write(asv_fasta, "ASVs.fa")

# count table:
asv_tab_standard <- as.matrix(asv_tab)
row.names(asv_tab_standard) <- sub(">", "", asv_headers)

# tax table:
asv_tax_standard <- as.matrix(asv_tax)
row.names(asv_tax_standard) <- sub(">", "", asv_headers)

# table asv name - sequence
table_asv_sequence <- cbind(asv_tax_standard[,0], rownames(asv_tax))
colnames(table_asv_sequence) <- c("Sequence")
write(table_asv_sequence, "ASVs_name_sequence.fa")

```

## Create phyloseq object
```{r}
# Phyloseq object with STANDARD
OTU_std = otu_table(as.matrix(asv_tab_standard), taxa_are_rows =TRUE)
TAX_std = tax_table(as.matrix(asv_tax_standard))
SAM = sample_data(metadata)

ps_std <- phyloseq(OTU_std, TAX_std, SAM)
ps_std1 <- subset_samples(ps_std, Species!="CuT" & Species!="CuP" & Species!="CuN" & Species!="CuG") # remove cullicoides	

asv_tab_std0 <- as(otu_table(ps_std1),"matrix")
write.table(asv_tab_std0, "0.asv_tab_standard0.tsv", sep="\t", quote=F, col.names=NA)

ps_std <- subset_samples(ps_std1, Sample!="NP16" & Sample!="NP17" & Sample!="NP18" & Sample!="NP19" & Sample!="S69" & Sample!="S70" & Sample!="S81" & Sample!="S82" & Sample!="NP20" & Sample!="NP21") # 11 juin 2019
ps_std <- prune_taxa(taxa_sums(ps_std) >= 1, ps_std) # remove asv that are not present in samples 
ps_std <- prune_samples(sample_sums(ps_std) >= 1, ps_std) # remove sample with 0 read


# Phyloseq object WITHOUT STANDARD
OTU = otu_table(asv_tab, taxa_are_rows =TRUE)
TAX = tax_table(as.matrix(asv_tax))
TREE = phy_tree(tree)

ps <- phyloseq(OTU, TAX, SAM, TREE)
ps <- subset_samples(ps, Species!="CuT" & Species!="CuP" & Species!="CuN" & Species!="CuG") # remove cullicoides
ps <- subset_samples(ps, Sample!="NP16" & Sample!="NP17" & Sample!="NP18" & Sample!="NP19" & Sample!="S69" & Sample!="S70" & Sample!="S81" & Sample!="S82" & Sample!="NP20" & Sample!="NP21")
ps <- prune_taxa(taxa_sums(ps) >= 1, ps) # remove asv that are not present in samples
ps <- prune_samples(sample_sums(ps) >= 1, ps) # remove counts = 0

write.table(asv_tab_standard, "1.asv_tab_standard.tsv", sep="\t", quote=F, col.names=NA)

ps
```

## Decontamination 
```{r}
setwd(path_decontam)

######################################## PREPROCESS #########################################

df <- as.data.frame(sample_data(ps)) # Put sample_data into a ggplot-friendly data.frame
df$LibrarySize <- sample_sums(ps)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))

pdf("LibrarySize.pdf")
ggplot(data=df, aes(x=Index, y=LibrarySize, color=Dna)) + geom_point()
dev.off()

as.numeric(get_variable(ps, "Dna"))
get_variable(ps, "Dna")
sample_data(ps)

sample_data(ps)$Dna <- as.numeric(get_variable(ps, "Dna"))



######################  IDENTIFY CONTAMINANTS - PREVALENCE #################################

sample_data(ps)$is.neg <- sample_data(ps)$Control == "Control sample"
sample_data(ps_std)$is.neg <- sample_data(ps_std)$Control == "Control sample"

contamdf.prev01 <- isContaminant(ps, method="prevalence", neg="is.neg", threshold=0.1)
contamdf.prev01_std <- isContaminant(ps_std, method="prevalence", neg="is.neg", threshold=0.1)
table(contamdf.prev01$contaminant) # 36 contaminants

contam_asvs_prev01 <- row.names(contamdf.prev01[contamdf.prev01$contaminant == TRUE, ])
contam_asvs_prev01_std <- row.names(contamdf.prev01_std[contamdf.prev01_std$contaminant == TRUE, ])

contam_standard <-asv_tax_standard[row.names(asv_tax_standard) %in% contam_asvs_prev01_std, ]
contam <-asv_tax[row.names(asv_tax) %in% contam_asvs_prev01, ]


ps.pa <- transform_sample_counts(ps, function(abund) 1*(abund>0))
ps.pa.neg <- prune_samples(sample_data(ps.pa)$Control == "Control sample", ps.pa)
ps.pa.pos <- prune_samples(sample_data(ps.pa)$Control == "True sample", ps.pa)


# Make data.frame of prevalence in positive and negative samples
df.pa <- data.frame(pa.pos=taxa_sums(ps.pa.pos), pa.neg=taxa_sums(ps.pa.neg),
                    contaminant=contamdf.prev01$contaminant)

pdf("Prevalence_Decontam.pdf")
ggplot(data=df.pa, aes(x=pa.neg, y=pa.pos, color=contaminant)) + geom_point() +
  xlab("Prevalence (Negative Controls)") + ylab("Prevalence (True Samples)")
dev.off()

# write table of contaminant asv 
write.table(contam, "asv_contam_prev01.tsv",
            sep="\t", quote=F, col.names=NA)
```

## New tables 
```{r}
# new otu table 
asv_tab_prev01 <- asv_tab[!row.names(asv_tab) %in% contam_asvs_prev01, ]
asv_tab_prev01_std <- asv_tab_standard[!row.names(asv_tab_standard) %in% contam_asvs_prev01_std, ]

# new tax table
asv_tax_prev01 <- asv_tax[!row.names(asv_tax) %in% contam_asvs_prev01, ]
asv_tax_prev01_std <- asv_tax_standard[!row.names(asv_tax_standard) %in% contam_asvs_prev01_std, ]

# new fasta  
contam_indices <- which(asv_fasta %in% paste0(">", contam_asvs_prev01))
dont_want <- sort(c(contam_indices, contam_indices + 1))
asv_fasta_prev01 <- asv_fasta[- dont_want]

# write tables
write.table(asv_tab_prev01, "asv_tab_prev01.tsv", sep="\t", quote=F, col.names=NA)
write.table(asv_tab_prev01_std, "2.asv_tab_prev01_std.tsv", sep="\t", quote=F, col.names=NA)

write.table(asv_tax_prev01, "asv_tax_prev01.tsv", sep="\t", quote=F, col.names=NA)
write.table(asv_tax_prev01_std, "asv_tax_prev01_std.tsv", sep="\t", quote=F, col.names=NA)

write(asv_fasta_prev01, "fasta_prev01.fa")


# extract new metadata as dataframe from phyloseq 
meta <- function(x) { # https://rdrr.io/github/microbiome/microbiome/src/R/meta.R
  df <- as(sample_data(x), "data.frame")
  rownames(df) <- sample_names(x)
  df
}
metadata_prev01 <- meta(ps)

# write decontamed metadata
write.table(metadata_prev01, "metadata_decontam.tsv", sep="\t", quote=F, col.names=NA)
```

## New decontam phyloseq object
```{r}
OTU_decontam <- otu_table(asv_tab_prev01_std, taxa_are_rows = TRUE)
TAX_decontam <- tax_table(as.matrix(asv_tax_prev01_std))
SAM_decontam <- sample_data(metadata_prev01)

ps_decontam <- phyloseq(OTU_decontam, TAX_decontam, SAM_decontam)
ps_decontam <- prune_taxa(taxa_sums(ps_decontam) >= 1, ps_decontam)# remove counts = 0 in tax_table
ps_decontam
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 2418 taxa and 208 samples ]
# sample_data() Sample Data:       [ 208 samples by 16 sample variables ]
# tax_table()   Taxonomy Table:    [ 2418 taxa by 7 taxonomic ranks ]

ps_decontam <- subset_taxa(ps_decontam, Kingdom == "Bacteria") # select only Bacteria
ps_decontam <- prune_samples(sample_sums(ps_decontam) >= 1, ps_decontam) # remove counts = 0 in otu_table

ps_decontam <- subset_samples(ps_decontam, Control!="Control sample") # remove blanks
```

## Removing remaining choloroplast and mitochondria
```{r}
ps_decontam2 <- subset_taxa(ps_decontam, Family!="Mitochondria" & Order!="Chloroplast")
ps_decontam2 <- prune_samples(sample_sums(ps_decontam2) >= 1, ps_decontam2)
ps_decontam2 <- prune_taxa(taxa_sums(ps_decontam2) >= 1, ps_decontam2)

asv_decontam2 <- as(otu_table(ps_decontam2),"matrix")

write.table(asv_decontam2, "3.asv_decontam2.tsv",sep="\t", quote=F, col.names=NA)

mitochondria <- subset_taxa(ps_decontam, Family=="Mitochondria") # 1 mitochondria 
chloroplast <- subset_taxa(ps_decontam, Order=="Chloroplast") # 26 chloroplast 

tax_mitochondria <- as(tax_table(mitochondria),"matrix")
tax_chloroplast <- as(tax_table(chloroplast),"matrix")
```

## Normalization 
```{r}
# % 
ps_percent <- transform_sample_counts( ps_decontam2, function(x) x/sum(x)*100 )
asv_percent <- as(otu_table(ps_percent),"matrix")
write.table(asv_percent, "4.asv_percent.tsv", sep="\t", quote=F, col.names=NA)


# deseq 
ps_deseq <- transform_sample_counts(ps_decontam2, function(x) x+1)

otu_deseq <- as(otu_table(ps_deseq),"matrix")
metadata_deseq <- as(sample_data(ps_deseq),"matrix")

deseq_counts <- DESeqDataSetFromMatrix(otu_deseq, colData = metadata_deseq, design = ~Dna)
deseq_counts_vst <- varianceStabilizingTransformation(deseq_counts)

asv_deseq <- as.matrix(assay(deseq_counts_vst))

write.table(asv_deseq, "5.asv_deseq.tsv", sep="\t", quote=F, col.names=NA)

otu_table(ps_deseq) <- otu_table(asv_deseq, taxa_are_rows = TRUE)

ps_deseq
```

## Exploring Wolbachia and Proteobacteria
```{r}
ps_wolbachia <- subset_taxa(ps_percent, Genus == "Wolbachia")
ps_wolbachia <- prune_taxa(taxa_sums(ps_wolbachia) >= 1, ps_wolbachia)
ps_wolbachia <- prune_samples(sample_sums(ps_wolbachia) >= 1, ps_wolbachia)
ps_wolbachia



ps_proteo <- subset_taxa(ps_percent, Phylum=="Proteobacteria")
ps_proteo <- prune_taxa(taxa_sums(ps_proteo) >= 1, ps_proteo)
ps_proteo <- prune_samples(sample_sums(ps_proteo) >= 1, ps_proteo)
ps_proteo
```

## Save phyloseq objects
```{r}
setwd(path_main)
save(ps, ps_decontam, ps_decontam2, ps_percent, ps_wolbachia, ps_proteo, ps_deseq, file = "objects.RData")

setwd(path)
save(ps, ps_decontam, ps_decontam2, ps_percent, ps_wolbachia, ps_proteo, ps_deseq, file = "objects.RData")

```

## Save session
```{r}
#install.packages("session", repos = "http://cran.us.r-project.org")
library(session); packageVersion("session")
save.session("metabarcoding2.Rda")
#load("phyloseq.Rda")
```

# III : Exploring data 

## Set path and load packages
```{r}
setwd(path)

dir.create("plot") # folder for plot
path2 <- "D:/stage/data/runs_new2/plot"



library("phyloseq")
library("DESeq2")
library("plotly")
library("hrbrthemes")
    
library("vegan")
library("ggplot2")
library("dendextend")
library("tidyr")
library("viridis")
library("reshape")
library("grid")
library("plyr")

scripts <- c("graphical_methods.R",
             "tree_methods.R",
             "plot_merged_trees.R",
             "specificity_methods.R",
             "ternary_plot.R",
             "richness.R",
             "edgePCA.R",
             "copy_number_correction.R",
             "import_frogs.R",
             "prevalence.R",
             "compute_niche.R",
             "NCM_fit.R")
urls <-
  paste0("https://raw.githubusercontent.com/mahendra-mariadassou/phyloseq-extended/master/R/",
         scripts)

# you need a connection internet or a local script copy
for (url in urls) {
  source(url)
}
```
## Distribution
```{r}
readsumsdf <- data.frame(nreads = sort(taxa_sums(ps), TRUE),
                         sorted = 1:ntaxa(ps), 
                         type = "OTU")

ggplot(readsumsdf, 
       aes(x = sorted, y = nreads)) + 
  geom_bar(stat = "identity") + 
  scale_y_log10() 

readsumsdf2 <- data.frame(nreads = sort(sample_sums(ps), TRUE), 
                          sorted = 1:nsamples(ps), 
                          type = "Samples")

readsumsdf3 <- rbind(readsumsdf,readsumsdf2)

p  <-  ggplot(readsumsdf3, 
              aes(x = sorted, y = nreads)) + 
  geom_bar(stat = "identity")+
  theme_gray()

p + ggtitle("Total number of reads before Preprocessing") + scale_y_log10() + facet_wrap(~type, 1, scales = "free")
```
## Change order of columns
```{r}
sample_data(ps_decontam2)$Organ <- factor(sample_data(ps_decontam2)$Organ, 
                                          levels=c("Whole", "Pool", "Intestine", "Ovary", "Salivary gland"))

sample_data(ps_percent)$Organ <- factor(sample_data(ps_percent)$Organ, 
                                          levels=c("Whole", "Pool", "Intestine", "Ovary", "Salivary gland"))

sample_data(ps_proteo)$Organ <- factor(sample_data(ps_proteo)$Organ, 
                                        levels=c("Whole", "Pool", "Intestine", "Ovary", "Salivary gland"))

sample_data(ps_proteo)$Species <- factor(sample_data(ps_proteo)$Species, 
                                       levels=c("Culex pipiens", "Culex quinquefasciatus", "Aedes aegypti"))
```

## Rarefaction curve 
```{r}
ps_nopool <- subset_samples(ps_decontam2, Organ!="Pool")

p1 <- ggrare(ps_nopool,
             step = 500,
             #color = "red",
             plot = T,
             parallel = F,
             se = T)

p2 <- p1 + 
  facet_wrap(~ Organ) + 
  geom_vline(xintercept = min(sample_sums(ps)), 
             color = "gray60") +
  xlim(0,100000) +
  ylim(0, 230) +
  labs(title = "Suffisant observations have been made for sampling",
       caption = "Rarefaction curve",
       x = "Sample Size", y = "Species Richness")


plot(p2)

```
## Alpha diversity 
```{r}
# Wolbachia - 
ps_whole <- subset_samples(ps_decontam2, Organ=="Whole")
ps_whole <- subset_samples(ps_whole, Organ!="Pool")


data4 <-  filter_taxa(ps_whole, 
                      function(x) sum(x >= 10) > (1), 
                      prune =  TRUE) 

p1 <- plot_richness(data4, 
                    x="Sample", 
                    color="Location", 
                    measures=c("Observed","Shannon", "Chao1"), 
                    nrow = 1) +
  ggtitle("Alpha diversity")

ggplot(p1$data,aes(Organ,value,colour=Location)) +
  facet_grid(variable ~ Species, drop=T,scale="free",space="fixed") +
  geom_boxplot(outlier.colour = NA,alpha=0.8, 
               position = position_dodge(width=0.9)) + 
  geom_point(size=2,position=position_jitterdodge(dodge.width=0.9)) +
  labs(title = expression(paste("A greater diversity for Aedes and lab samples")),
       caption = "Alpha diversity", y = "Diversity index")





ps_aedes <- subset_samples(ps_decontam2, Species!="Aedes aegypti" & Organ!="Pool")

data5 <-  filter_taxa(ps_aedes, 
                      function(x) sum(x >= 10) > (1), 
                      prune =  TRUE) 

p2 <- plot_richness(data5, 
                    x="Sample", 
                    color="Location", 
                    measures=c("Observed","Shannon", "Chao1"), 
                    nrow = 1) +
  ggtitle("Alpha diversity")


ggplot(p2$data,aes(Field,value,colour=Location,shape=Field)) +
  facet_grid(variable ~ Organ+Species, drop=T,scale="free",space="fixed") +
  geom_boxplot(outlier.colour = NA,alpha=0.8,
               position = position_dodge(width=0.9)) +
  geom_point(size=2,position=position_jitterdodge(dodge.width=0.9)) +
  labs(title = "A greater diversity in whole organisms",
       caption = "Alpha diversity", y = "Diversity index")

```

## Taxonomic composition 
```{r}

# Proteobacteria

ps_proteo_nopool <- subset_samples(ps_proteo, Organ!="Pool")
ps_proteo_nopool

p <- plot_composition(ps_proteo_nopool,
                      taxaRank1 = "Kingdom",
                      taxaSet1 ="Bacteria",
                      taxaRank2 = "Genus", 
                      numberOfTaxa = 20, 
                      fill= "Genus") +
  facet_wrap(~ Species+Organ+Location, scales = "free_x", ncol=5) + 
  theme(plot.title = element_text(hjust = 0.5), strip.text = element_text(size=8), strip.text.x = element_text(margin = margin(0.1,0,0.1,0, "cm"))) + 
  labs(title = "Wolbachia is the dominant genus",
       caption = "Taxonomic composition (20 most abundant genus)", x="Sample", y = "Abundance")

plot(p)

pdf("7-taxo_20_genus.pdf")
plot(p)
dev.off()


# Proteobacteria within the 20 most abundant genus

ps_proteo_nopool2 <- subset_taxa(ps_proteo_nopool, Genus!="Wolbachia" & Genus!="Acinetobacter" & Genus!="Aeromonas" & Genus!="Asaia" &
                                    Genus!="Enhydrobacter" & Genus!="Erwinia" & Genus!="Haemophilus" & Genus!="Klebsiella" & Genus!="Legionella" &
                                    Genus!="Massilia" & Genus!="Morganella" & Genus!="Providencia" & Genus!="Pseudomonas" & Genus!="Rahnella" &
                                    Genus!="Ralstonia" & Genus!="Serratia" & Genus!="Sphingomonas" & Genus!="Thorsellia" & Genus!="Zymobacter")
ps_proteo_nopool2

p <- plot_composition(ps_proteo_nopool2,
                      taxaRank1 = "Kingdom",
                      taxaSet1 ="Bacteria",
                      taxaRank2 = "Genus", 
                      numberOfTaxa = 20, 
                      fill= "Genus") +
  facet_wrap(~ Species+Organ+Location, scales = "free_x", ncol=5) + 
  theme(plot.title = element_text(hjust = 0.5), strip.text = element_text(size=8), strip.text.x = element_text(margin = margin(0.1,0,0.1,0, "cm"))) + 
  labs(title = "When removing the 20 most abundant genus",
       caption = "Taxonomic composition", x="Sample", y = "Abundance")

plot(p)

pdf("7bis-taxo_wt_20_genus.pdf")
plot(p)
dev.off()

```
## Beta diversity (NMDS)
```{r}

# Whole individuals 

ps_percent_whole <- subset_samples(ps_percent, Organ == "Whole")
ps_culex_whole <- subset_samples(ps_percent_whole, Species!="Aedes aegypti") # whole Culex

prop.percent_whole <- transform_sample_counts(ps_percent_whole, function(count_tab) count_tab/sum(count_tab))
bray.percent_whole <- ordinate(ps_percent_whole, method="NMDS", distance="bray")

prop.culex_whole <- transform_sample_counts(ps_culex_whole, function(count_tab) count_tab/sum(count_tab))
bray.culex_whole <- ordinate(ps_culex_whole, method="NMDS", distance="bray")

#pdf("7-NMDS_bray_whole.pdf")
plot_ordination(prop.percent_whole, bray.percent_whole, color="Species", title="Bray NMDS with Whole body", label="Sample") +
  labs(title = "Does species influence bacterial community structure ? ",
       caption = "Bray NMDS on whole mosquitoes", x="NMDS1", y = "NMDS2")+
  geom_point(size = 4) +
  theme_gray()
#dev.off()

#pdf("9-NMDS_bray_culex_whole_ellipse.pdf")
  plot_ordination(prop.culex_whole, bray.culex_whole, color="Field", shape="Location", title="Bray NMDS with Whole body - Location without Aedes aegypti", label="Sample") +
  labs(title = "Do antibiotics influence microbiote ? ",
       caption = expression(paste("Bray NMDS on whole ", italic("Culex")," whole mosquitoes")), x="NMDS1", y = "NMDS2") +
  stat_ellipse(geom = "polygon", level=0.70,alpha = 1/2, aes(fill = Species))+
  scale_fill_manual(values=c("yellow","green"))+
  scale_color_manual(values=c("red", "blue"))+
  geom_point(size = 5) +
  theme_gray()
#dev.off()

```

```{r}
# Organs

ps_pipiens_organ <- subset_samples(ps_percent, Organ!="Whole" & Species=="Culex pipiens") # organ samples of Culex pipiens
ps_pipiens_organ <- subset_samples(ps_pipiens_organ, Sample!="S175") # removing outlier
ps_pipiens_organ_camping <- subset_samples(ps_pipiens_organ, Location=="Camping Europe") # organ samples of Culex pipiens from Camping Europe

prop.pipiens_organ_camping <- transform_sample_counts(ps_pipiens_organ_camping, function(count_tab) count_tab/sum(count_tab))
bray.pipiens_organ_camping <- ordinate(ps_pipiens_organ_camping, method="NMDS", distance="bray")

#pdf("11-NMDS_bray_organs_culex_CE.pdf")
plot_ordination(prop.pipiens_organ_camping, bray.pipiens_organ_camping, color="Organ", title="Bray NMDS", label="Sample") +
  labs(title = expression(paste("Organ influences the structure of ", italic("Culex pipiens"), " bacterial community")),
       caption = expression(paste("Bray NMDS on ovary, intestine and salivary gland of ", italic('Culex pipiens'), " from Camping Europe")), x="NMDS1", y = "NMDS2")+
  geom_point(size = 4) +
  theme_gray()
#dev.off()

```

```{r}
# Time (intestine)

ps_intestine_camping <- subset_samples(ps_percent, Location == "Camping Europe" & Organ =="Intestine") # intestine samples from Camping Europe
ps_intestine_camping_date <- subset_samples(ps_intestine_camping, Date =="30/05/2017" | Date =="28/06/2017") # intestine samples from Camping Europe extracted the 30/05/2017 and 28/06/2017

prop.intestine_camping_date <- transform_sample_counts(ps_intestine_camping_date, function(count_tab) count_tab/sum(count_tab))
bray.intestine_camping_date <- ordinate(ps_intestine_camping_date, method="NMDS", distance="bray")

#pdf("13-NMDS_bray_intestine_CE_date.pdf")
plot_ordination(prop.intestine_camping_date, bray.intestine_camping_date, color="Date", title="Bray NMDS with intestine of Culex pipiens in Camping Europe", label="Sample")+
  labs(title = expression(paste("Does time of sampling influence the microbiote of ", italic("Culex pipiens"),"?")),
       caption = "Bray NMDS on intestine at Camping Europe", x="NMDS1", y = "NMDS2") +
  geom_point(size = 4) +
  theme_gray()
#dev.off()


# Time (ovary)

ps_percent_ovary <- subset_samples(ps_percent, Organ == "Ovary")
ps_ovary_culex_camping <- subset_samples(ps_percent_ovary, Species=="Culex pipiens" & Location == "Camping Europe") # ovary samples from Camping Europe
ps_ovary_culex_camping_date <- subset_samples(ps_ovary_culex_camping, Date =="30/05/2017" | Date =="28/06/2017") # samples extracted the 30/05/2017 and 28/06/2017
ps_ovary_culex_camping_date <- subset_samples(ps_ovary_culex_camping_date, Sample!="S103") # removing outlier

prop.ovary_culex_camping_date <- transform_sample_counts(ps_ovary_culex_camping_date, function(count_tab) count_tab/sum(count_tab))
bray.ovary_culex_camping_date <- ordinate(ps_ovary_culex_camping_date, method="NMDS", distance="bray")

#pdf("14-NMDS_bray_ovary_CE_date.pdf")
plot_ordination(prop.ovary_culex_camping_date, bray.ovary_culex_camping_date, color="Date", title="Bray NMDS with ovary - Culex, Camping Europe, Dates", label="Sample") +
  labs(title = expression(paste("Does time of sampling influence the microbiote of ", italic("Culex pipiens"),"?")),
       caption = "Bray NMDS on intestine at Camping Europe", x="NMDS1", y = "NMDS2") +
  geom_point(size = 4) +
  theme_gray()
#dev.off()


```

## Heatmap 

```{r}
# Heatmap within Culex pipiens samples

ps_pipiens <- subset_samples(ps_percent, Species=="Culex pipiens")
ps_pipiens_wolbachia <- subset_taxa(ps_pipiens, Genus=="Wolbachia") # Culex pipiens samples with Wolbachia
ps_pipiens_wolbachia <- prune_taxa(taxa_sums(ps_pipiens_wolbachia) >= 1, ps_pipiens_wolbachia) # removing samples without taxa
ps_pipiens_wolbachia <- prune_samples(sample_sums(ps_pipiens_wolbachia) >= 1, ps_pipiens_wolbachia) # removing samples without read
ps_pipiens_wolbachia <- prune_taxa(names(sort(taxa_sums(ps_pipiens_wolbachia),TRUE)[1:30]), ps_pipiens_wolbachia) # select the 30 most abundant Wolbachia samples

# heatmap with transformed abundance (log10)
plot_heatmap(ps_pipiens_wolbachia, sample.label="Location", sample.order="Location", low="#000033", high="#00FF00", trans=log10_trans())+
  facet_wrap(~ Field + Organ, scales = "free_x", ncol = 4)+
  labs(title = expression(paste("ASV1 is abundant in almost all sequences of ", italic("Culex pipiens"))),
       caption = expression(paste("Heatmap of ", italic('Culex pipiens'), " that contains Wolbachia (log10 transformation)")), x="Field", y = "ASV")


```

```{r}
# Heatmap within Culex quinquefasciatus samples

ps_quinque <- subset_samples(ps_decontam2, Species=="Culex quinquefasciatus")
ps_quinque_wolbachia <- subset_taxa(ps_quinque, Genus=="Wolbachia") # Culex quinquefasciatus samples with Wolbachia
ps_quinque_wolbachia <- prune_taxa(taxa_sums(ps_quinque_wolbachia) >= 1, ps_quinque_wolbachia) # removing samples without taxa
ps_quinque_wolbachia <- prune_samples(sample_sums(ps_quinque_wolbachia) >= 1, ps_quinque_wolbachia) # removing samples without read
ps_quinque_wolbachia <- prune_taxa(names(sort(taxa_sums(ps_quinque_wolbachia),TRUE)[1:30]), ps_quinque_wolbachia) # select the 30 most abundant Wolbachia samples

# heatmap with transformed abundance (log10)
plot_heatmap(ps_quinque_wolbachia, sample.label="Location", sample.order="Location", low="#000033", high="#00FF00",trans=log10_trans())+
  facet_wrap(~ Organ, scales = "free_x", ncol = 3)+
  labs(title = expression(paste("ASV1 is abundant in almost all sequences of ", italic("Culex quinquefasciatus"), ", especially in ovary and pool samples")),
       caption = expression(paste("Heatmap of ", italic('Culex quinquefasciatus'), " that contains Wolbachia (log10 transformation)")), x="Field", y = "ASV")

```

```{r}
# Heatmap within Aedes aegypti samples

ps_aedes <- subset_samples(ps_decontam2, Species=="Aedes aegypti" & Organ!="Pool")
ps_aedes_wolbachia <- subset_taxa(ps_aedes, Genus=="Wolbachia") # Aedes aegypti samples with Wolbachia
ps_aedes_wolbachia <- prune_taxa(taxa_sums(ps_aedes_wolbachia) >= 1, ps_aedes_wolbachia) # removing samples without taxa
ps_aedes_wolbachia <- prune_samples(sample_sums(ps_aedes_wolbachia) >= 1, ps_aedes_wolbachia) # removing samples without read
ps_aedes_wolbachia <- prune_taxa(names(sort(taxa_sums(ps_aedes_wolbachia),TRUE)[1:30]), ps_aedes_wolbachia) # select the 30 most abundant Wolbachia samples

# heatmap with transformed abundance (log10)
plot_heatmap(ps_aedes_wolbachia, sample.label="Location", sample.order="Location", low="#000033", high="#00FF00", trans=log10_trans())+
  facet_wrap(~ Organ, scales = "free_x", ncol = 2)+
  labs(title = expression(paste("ASV1 is abundant in almost all sequences of ", italic("Aedes aegypti"),", especially in ovary samples")),
       caption = expression(paste("Heatmap of  ", italic('Aedes aegypti'), " that contains Wolbachia (log10 transformation)")), x="Field", y = "ASV")



```


## Statistical tests 

### Difference between field and labo pipiens samples microbiota ? (field influence)

```{r}
# Permanova

ps_pipiens_whole <- subset_samples(ps_percent_whole, Species=="Culex pipiens")
ps_pipiens_whole_bosc_lavar <- subset_samples(ps_pipiens_whole, Location!="Camping Europe") # whole Culex pipiens from Lavar and Bosc

adonis(vegdist(t(otu_table(ps_pipiens_whole_bosc_lavar)), method = "bray") ~Field,
       data=as(sample_data(ps_pipiens_whole_bosc_lavar), "data.frame"), permutation = 9999)

ps_pipiens_whole_camping_lavar <- subset_samples(ps_pipiens_whole, Location!="Bosc") # whole Culex pipiens from Lavar and Camping

adonis(vegdist(t(otu_table(ps_pipiens_whole_camping_lavar)), method = "bray") ~Location,
       data=as(sample_data(ps_pipiens_whole_camping_lavar), "data.frame"), permutation = 9999)

```

```{r}
# Anosim

meta_pipiens_bosc_lavar <- sample_data(ps_pipiens_whole_bosc_lavar)

anosim(vegdist(t(otu_table(ps_pipiens_whole_bosc_lavar))), meta_pipiens_bosc_lavar$Location, permutations=1000)

meta_pipiens_camping_lavar <- sample_data(ps_pipiens_whole_camping_lavar)

anosim(vegdist(t(otu_table(ps_pipiens_whole_camping_lavar))), meta_pipiens_camping_lavar$Location, permutations=1000)

```

### Difference between whole Aedes aegypti and Culex quinquefasciatus microbiota from Guadeloupe ? (species influence)

```{r}
# Permanova
ps_gwada <- subset_samples(ps_percent, Location=="Guadeloupe" | Field=="Field")
ps_whole_gwada <- subset_samples(ps_gwada, Organ=="Whole")

adonis(vegdist(t(otu_table(ps_whole_gwada)), method = "bray") ~Species,
       data=as(sample_data(ps_whole_gwada), "data.frame"), permutation = 9999)

```

```{r}
# Anosim 
meta_whole_gwada <- sample_data(ps_whole_gwada)

anosim(vegdist(t(otu_table(ps_whole_gwada))), meta_whole_gwada$Species, permutations=1000)

```

### Difference between whole field and wolbachia- Culex quinquefasciatus microbiota ? (antibiotics influence)

```{r}
# Permanova
ps_quinque_whole <- subset_samples(ps_quinque, Organ=="Whole")

adonis(vegdist(t(otu_table(ps_quinque_whole)), method = "bray") ~Field,
       data=as(sample_data(ps_quinque_whole), "data.frame"), permutation = 9999)

```

```{r}
# Anosim 
meta_quinque_whole <- sample_data(ps_quinque_whole)

anosim(vegdist(t(otu_table(ps_quinque_whole))), meta_quinque_whole$Field, permutations=1000)

```


### Does location influence the structure of whole Culex microbiote ? (location influence)

```{r}
# Permanova
ps_culex_whole_field <- subset_samples(ps_culex_whole, Location!="Labo Tetracycline" & Location!="Lavar")

adonis(vegdist(t(otu_table(ps_culex_whole_field)), method = "bray") ~Location,
       data=as(sample_data(ps_culex_whole_field), "data.frame"), permutation = 9999)

```

```{r}
# Anosim 
meta_culex_whole_field <- sample_data(ps_culex_whole_field)

anosim(vegdist(t(otu_table(ps_culex_whole_field))), meta_culex_whole_field$Location, permutations=1000)

```


### Difference between culex samples from Camping Europe microbiota depends on two dates ? (date influence)

```{r}
# Permanova
ps_pipiens_camping <- subset_samples(ps_pipiens, Location=="Camping Europe")
ps_pipiens_ovary_camping <- subset_samples(ps_pipiens_camping, Organ=="Ovary")

adonis(vegdist(t(otu_table(ps_pipiens_ovary_camping)), method = "bray") ~ Date,
       data=as(sample_data(ps_pipiens_ovary_camping), "data.frame"), permutation = 9999)

ps_pipiens_intestine_camping <- subset_samples(ps_pipiens_camping, Organ=="Intestine")

adonis(vegdist(t(otu_table(ps_pipiens_intestine_camping)), method = "bray") ~ Date,
       data=as(sample_data(ps_pipiens_intestine_camping), "data.frame"), permutation = 9999)

```

```{r}
# Anosim 
meta_pipiens_ovary_camping <- sample_data(ps_pipiens_ovary_camping)
anosim(vegdist(t(otu_table(ps_pipiens_ovary_camping))), meta_pipiens_ovary_camping$Date, permutations=1000)

meta_pipiens_intestine_camping <- sample_data(ps_pipiens_intestine_camping)
anosim(vegdist(t(otu_table(ps_pipiens_intestine_camping))), meta_pipiens_intestine_camping$Date, permutations=1000)
```


### Difference between culex pipiens samples microbiota depends on organs ? (organs influence)

```{r}
# Permanova
adonis(vegdist(t(otu_table(ps_pipiens_organ)), method = "bray") ~ Organ,
       data=as(sample_data(ps_pipiens_organ), "data.frame"), permutation = 9999)

```

```{r}
# Anosim 
meta_pipiens_organ <- sample_data(ps_pipiens_organ)
anosim(vegdist(t(otu_table(ps_pipiens_organ))), meta_pipiens_organ$Organ, permutations=1000)

```

```{r}
# Pairwise permanova

#install.packages("devtools")
#devtools::install_github("leffj/mctoolsr")

library("mctoolsr")

calc_pairwise_permanovas(as(vegdist(t(otu_table(ps_pipiens_organ))), "matrix"), as(sample_data(ps_pipiens_organ), "data.frame"), "Organ")

```






# Save session 
```{r}
library(session); packageVersion("session")
save.session("metabarcoding_workflow.Rda")
```


